src/python/pants/rules/core/run.py

```diff
@@ -49,12 +50,61 @@ async def run(
     workspace: Workspace,
     runner: InteractiveRunner,
     build_root: BuildRoot,
-    addresses: Addresses,
+    targets: Targets,
     options: RunOptions,
     global_options: GlobalOptions,
+    union_membership: UnionMembership,
+    registered_target_types: RegisteredTargetTypes,
 ) -> Run:
-    address = addresses.expect_single()
-    binary = await Get[CreatedBinary](Address, address)
+    config_types: Iterable[Type[BinaryConfiguration]] = union_membership.union_rules[
+        BinaryConfiguration
+    ]
+    valid_config_types_by_target: Dict[Target, List[Type[BinaryConfiguration]]] = {}
+    for target in targets:
+        for config_type in config_types:
+            if config_type.is_valid(target):
+                valid_config_types_by_target.setdefault(target, []).append(config_type)
+
+    if not valid_config_types_by_target:
+        all_valid_target_types = itertools.chain.from_iterable(
+            config_type.valid_target_types(
+                registered_target_types.types, union_membership=union_membership
+            )
+            for config_type in config_types
+        )
+        valid_target_type_aliases = sorted(
+            target_type.alias for target_type in all_valid_target_types
+        )
+        raise ValueError(
+            f"The `run` goal only works with the following target types: "
+            f"{', '.join(valid_target_type_aliases)}\n\nYou used {target.address} with target type "
+            f"{target.alias}."
```

Comment:
I think `target` is a bad reference. It will refer to the last element of the iterable `targets`. I liked the message from `binary.py` and think you can copy it over here.

Reply:
Ouch - good catch. I factored this "at least one" checking up into a centralized helper.